x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportGrantAction.java

```diff
@@ -56,12 +55,6 @@ public final void doExecute(Task task, Request request, ActionListener<Response>
         try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
             final AuthenticationToken authenticationToken = request.getGrant().getAuthenticationToken();
             assert authenticationToken != null : "authentication token must not be null";
-            if (authenticationToken == null) {
```

Comment:
This is not needed anymore because `request.getGrant().getAuthenticationToken()` cannot return `null` even for an unrecognized grant types (which is also not possible in _practice_ because of request validation).
Instead `request.getGrant().getAuthenticationToken()` throws an exception for an unrecognized grant type (as well as for using `client_authentication` alongside regular ES access tokens), which is translated to a `listener.onFailure` call by the try-catch handler (which is tested to return 400 to REST clients).

Reply:
In other words the `request.getGrant().getAuthenticationToken()` used `null` to signal a error condition (unrecognized grant type), but now there's another error condition (client authentication used alongside regular ES access tokens), so I changed the method to throw exceptions in these 2 cases and never return `null`.